modules/issue_tracker/php/edit.class.inc

```diff
@@ -0,0 +1,860 @@
+            $this->showError("Visit Label must be accompanied by a PSCID");
+        }
+        return $result;
+    }
+
+    /**
+     * Iterates through submitted values and filters only values that have changed
+     *
+     * @param array $issueValues new values
+     * @param int   $issueID     issue ID
+     * @param \User $user        current user
+     *
+     * @return array array of changed values
+     */
+    function getChangedValues(array $issueValues, int $issueID, \User $user): array
+    {
+        $issueData     = $this->getIssueData($issueID, $user);
+        $changedValues = [];
+        foreach ($issueValues as $key => $value) {
+            // Only include fields that have changed
+            // centerID is allowed to be NULL
+            if ($issueValues[$key] != ($issueData[$key] ?? '')
+                && (!empty($value) || $key === 'centerID')
+            ) {
+                $changedValues[$key] = $value;
+            }
+        }
+        return $changedValues;
+    }
+
+    /**
+     * Puts updated fields into the issues_history table.
+     *
+     * @param array $values  the new values
+     * @param int   $issueID the issue ID
+     * @param \User $user    the user
+     *
+     * @return void
+     */
+    function updateHistory(array $values, int $issueID, \User $user)
+    {
+        $db = $this->loris->getDatabaseConnection();
+        foreach ($values as $key => $value) {
+            // centerID is allowed to be NULL
+            if (!empty($value) || $key === 'centerID') {
+                $changedValues = [
+                    'newValue'     => $value ?? '',
+                    'fieldChanged' => $key,
+                    'issueID'      => $issueID,
+                    'addedBy'      => $user->getUsername(),
+                ];
+                $db->insert('issues_history', $changedValues);
+            }
+        }
+    }
+
+    /**
+     * Puts updated fields into the issues_comments table.
+     *
+     * @param string $comment new issue comment
+     * @param int    $issueID the issue ID
+     * @param \User  $user    the user
+     *
+     * @return void
+     */
+    function updateComments(string $comment, int $issueID, \User $user)
+    {
+        $db = $this->loris->getDatabaseConnection();
+        if (isset($comment) && $comment != "null") {
+            $commentValues = [
+                'issueComment' => $comment,
+                'addedBy'      => $user->getUsername(),
+                'issueID'      => $issueID,
+            ];
+            $db->insert('issues_comments', $commentValues);
+        }
+    }
+
+    /**
+     * Utility function to return errors from the server
+     *
+     * @param string $message error message to display
+     *
+     * @return void
+     */
+    function showError(string $message) : void
+    {
+        if (!isset($message)) {
+            $message = 'An unknown error occurred!';
+        }
+        header('HTTP/1.1 400 Bad Request');
```

Comment:
These are side-effect inducing functions that aren't valid in an class that is supposed to be returning a PSR ResponseInterface.